### PR TITLE
Use Hashicorp Vault for system key when specified

### PR DESCRIPTION
**Pull request by mandoonandy**
**Discussion only:** *It can be merged but may not be the optimal solution*

Possible use of Hashicorp Vault. The (random or not) key should first be stored in the vault. Then deploys can proceed pulling that key.

The password is not randomly generated, since the ansible playbook would become non-idempotent. (Yes, the auth secret already does that... maybe fix that by checking the existence of the entry in sf.service) 

This is all presupposing we should not use an entirely different playbook for maintaining a production cluster.

**Setup**
```
pip install ansible-modules-hashivault    # Noted in error message of ansible
VAULT_SYSTEM_KEY_PATH=sfcluster/syd/system
```

### CODE REVIEW

File: ansible/deploy.yml [hunk 1]
```diff
@@ -328,8 +328,22 @@
         enabled: yes
         state: restarted
 
-    - name: Create an admin namespace called "system" with a password configured
-      shell: sf-passwd deploy "{{ADMIN_PASSWORD}}"
+    - name: Set admin password from extra-vars
+      set_fact:
+        system_key: "{{admin_password}}"
```
Comment (mikalstill) [inline ansible/deploy.yml]: Should this have a "when" so its either or? I suspect this way works, but I am unsure if its confusing or not.

File: ansible/deployandtest.sh [hunk 2]
```diff
@@ -138,9 +138,14 @@ ANSIBLE_VARS="$ANSIBLE_VARS cloud=$CLOUD"
 ANSIBLE_VARS="$ANSIBLE_VARS bootdelay=$BOOTDELAY"
 ANSIBLE_VARS="$ANSIBLE_VARS ansible_root=$cwd"
 ANSIBLE_VARS="$ANSIBLE_VARS uniqifier=$UNIQIFIER"
-ANSIBLE_VARS="$ANSIBLE_VARS ADMIN_PASSWORD=$ADMIN_PASSWORD"
+ANSIBLE_VARS="$ANSIBLE_VARS admin_password=$ADMIN_PASSWORD"
```
Comment (mikalstill) [inline ansible/deployandtest.sh:154]: Because if I am using Vault I wouldn't expect to have to set this variable, right?

**Review by mikalstill — CHANGES_REQUESTED** [review]
One question.

File: ansible/deploy.yml (hunk 1 above)
Comment (mandoonandy) [reply ansible/deploy.yml]: It looks verbose but that is ansible. It's really a variable with default initialisation followed by an if statement to set an alternative. 
I did have the same thought as you, but went for less lines of code to reduce bug surface area. Adding two separate if ("when") statements, you do have to be careful to be sure at least one of the "when" statements fire.
On the one hand, a future developer could misread intent with the one "when", on the other hand she could change a "when" such that neither (or both) fire.
Whoa. That got too long!

TL;DR
 I can see your position as well. Can add the "when" if you think its better.

File: ansible/deployandtest.sh (hunk 2 above)
Comment (mandoonandy) [reply ansible/deployandtest.sh:154]: Correct. If you don't set it, then a default will be set.
If you set the vault setting, then it is not used.
We could add a warning that you have set an admin AND set a vault password?

Comment (mikalstill) [reply ansible/deployandtest.sh:154]: I think that warning will one day save me from confusion, so let's do that. Thanks!

Comment (mandoonandy) [reply ansible/deployandtest.sh:154]: Add a full error to ensure the mistake is caught

**Comment by mandoonandy** [comment]
Found a few other issues. Fixed now.

**Review by mikalstill — APPROVED** [review]
LGTM

**Comment by mikalstill** [comment]
I don't think I will cherry-pick this one for 0.2.2 for now because it looks quite complicated. Is that ok?
